src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,287 @@
+
+use pulldown_cmark::{Parser, Tag};
+use std::io::{BufRead, BufReader};
+use std::process::Stdio;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    process::Command,
+};
+use std::{env, fs};
+
+use crate::dashboard;
+
+/// Parses the chapter/section hierarchy in the markdown file specified by
+/// `summary_path` and returns a mapping from markdown files containing rust
+/// code to corresponding directories where the extracted rust code should
+/// reside.
+fn parse_hierarchy(summary_path: &Path) -> HashMap<PathBuf, PathBuf> {
+    let summary = fs::read_to_string(summary_path).unwrap();
+    // Skip the title and introduction.
+    let parser = Parser::new(&summary).skip(8);
+    // Set "ref" as the root of the hierarchical path.
+    let mut hierarchy = PathBuf::from("ref");
+    let mut map = HashMap::new();
+    // Introduction is a especial case, so handle it separately.
+    map.insert(PathBuf::from("introduction.md"), hierarchy.join("Introduction"));
+    for event in parser {
+        match event {
+            pulldown_cmark::Event::End(Tag::Item) => {
+                // Pop the current chapter/section from the hierarchy once
+                // we are done processing it and its subsections.
+                hierarchy.pop();
+            }
+            pulldown_cmark::Event::End(Tag::Link(_, path, _)) => {
+                // At the start of the link tag, the hierarchy does not yet
+                // contain the title of the current chapter/section. So, we wait
+                // for the end of the link tag before adding the path and
+                // hierarchy of the current chapter/section to the map.
+                map.insert(path.split('/').collect(), hierarchy.clone());
+            }
+            pulldown_cmark::Event::Text(text) => {
+                // Add the current chapter/section title to the hierarchy.
+                hierarchy.push(text.to_string());
+            }
+            _ => (),
+        }
+    }
+    map
+}
+
+/// Extracts examples from the given relative `paths` in the `book_dir` and
+/// saves them in `gen_dir`.
+fn extract_examples(paths: Vec<&PathBuf>, book_dir: &Path, gen_dir: &Path) {
+    for path in paths {
+        let mut cmd = Command::new("rustdoc");
+        cmd.args([
+            "+nightly",
+            "--test",
+            "-Z",
+            "unstable-options",
+            book_dir.join(path).to_str().unwrap(),
+            "--test-builder",
+            &["src", "tools", "dashboard", "print.sh"]
+                .iter()
+                .collect::<PathBuf>()
+                .to_str()
+                .unwrap(),
+            "--persist-doctests",
+            gen_dir.to_str().unwrap(),
+            "--no-run",
+        ]);
+        cmd.stdout(Stdio::null());
+        cmd.spawn()
+            .expect(&format!("Error: failed to spawn process:\n{:?}", cmd))
+            .wait()
+            .expect(&format!("Error: spawned process did not run:\n{:?}", cmd));
```

Comment:
When can this happen?

Reply:
I don't know. I just followed the examples. But the error is too verbose and probably printed by default. Removed.